### ISSUE TITLE
Adding support for zstd

### ISSUE THREAD

**Issue by martinellimarco** [mxmlnkn/ratarmount#40]
Hi,
is it possible to add support for .tar.zst file made with [zstd](https://github.com/facebook/zstd) ?

I have a few big archives (hundreds of GB each) and the performance with archivemount are not good.

I tried converting a couple of archives in a more common .tar.bz2 but the resulted files are much bigger than the originals and I can't do that for all the archives.

It is my understanding that for this the first step is to write a new "indexed_zst" module. I tried looking at the code of indexed_bzip2 and indexed_gzip but I'm not a python programmer and I can't contribute much.

Is there anything I can do to support this format?

Regards

**Comment by mxmlnkn** [comment]
As you already found out, in principle, an `indexed_zst` python module would be sufficient to add support to ratarmount.

The already existing [zstandard](https://pypi.org/project/zstandard/) Python module does not seem to be sufficient because:
> The stream returned by stream_reader() is partially seekable. Absolute and relative positions (SEEK_SET and SEEK_CUR) forward of the current position are allowed. Offsets behind the current read position and offsets relative to the end of stream are not allowed and will raise ValueError if attempted. 
Backward seeking is not even supported, so it makes me think that forward seeking is also only emulated and simply decodes everything inbetween.

You don't have to be a Python programmer to contribute!

The actual meaty parts is basically a simple "seek" method. Everything else is mostly boiler-plate. In order to support a "seek" method in the first place, you would need a custom decoder though. That is the main problem. `indexed_bzip2` actually brings it's own complete BZ2 decoder. If possible, I'd want to avoid this for other formats. The [Zstd C-API](http://facebook.github.io/zstd/zstd_manual.html#Chapter3) seems to mention `blocks` and `frames` and also has some promisingly sounding functions like ` ZSTD_getFrameContentSize` and `ZSTD_findFrameCompressedSize`. So, maybe it's possible to use this API to hack together a decompressor, which remembers all frame compressed and content/uncompressed sizes and then is able to seek to the desired frame given an offset in the uncompressed data. If you could provide a proof of work for that even as a standalone C/C++ program, then the hardest work would be done.

**Comment by martinellimarco** [comment]
Got it, thanks for the quick reply and the excellent explanation. I'll try to hack something together this weekend.

**Comment by martinellimarco** [comment]
I've been experimenting a lot with zstd in the last few days.

To resume what I've found out:

A zst file produced with `zstd` cli tool is composed of a single frame, there is no header or trail. It's possible to seek forward (decompressing from the beginning) but to go backwards we need to start over.

There is an open issue about [limiting the frame size in `zstd`](https://github.com/facebook/zstd/issues/2121) and get a zst file composed of multiple frames.

Despite what the `zstd` cli tool procude, a zst file CAN be a concatenation of multiple frames and this can be produced easily with other tools (more on this below) or concatenating 2 .zst files together.
Use `zstd -l <file>.zst` to see the number of frames in a file.
A file decompressed this way is the concatenation of all the frames decompressed.

`ZSTD_findFrameCompressedSize` will always return the size of the compressed frame.

`ZSTD_getFrameContentSize` may return the size of the decompressed frame, but it can return ZSTD_CONTENTSIZE_UNKNOWN.
A frame has an unknown content size if the compressor don't know the size the moment it start writing the header.
This is the case if the compressor works with a stream of data.

For example:
```
dd if=/dev/zero bs=1M count=1 | zstd -o test.zst
zstd -l test.zst
```

results in

```
$ zstd -l test.zst
Frames  Skips  Compressed  Uncompressed  Ratio  Check  Filename
     1      0     0.05 KB                       XXH64  test.zst
```

As you can see the uncompressed size is unknown.
On the other hand

```
dd if=/dev/zero bs=1M count=1 of=test2
zstd test2
zstd -l test2.zst
```

results in

```
$ zstd -l test2.zst
Frames  Skips  Compressed  Uncompressed  Ratio  Check  Filename
     1      0     0.05 KB    1024.00 KB  19418.074  XXH64  test2.zst
```
and this time the uncompressed size is known.

It is also possible to have a zst file with mixed frames, some with known and some with unknown uncompressed size.
You can test it with

```
cat test.zst test2.zst > test3.zst
zstd -l test3.zst
```

that produces

```
$ zstd -l test3.zst 
Frames  Skips  Compressed  Uncompressed  Ratio  Check  Filename
     2      0     0.10 KB                       XXH64  test3.zst
```

Here it should be noted that the file is perfectly valid, it produces a 2MB uncompressed file and zstd report it as having 2 frames (and 0 skippable frames).

This limits the ability to build a seek table as we can easily build one for each frame that has known size from the beginning of the file but we have to stop as soon as we find a frame with unknown uncompressed size.
Alternatively one can decompress such frame to calculate the size and fill the skip table or the skip table.

I've found out from the issue linked above that there is a [seekable format](https://github.com/facebook/zstd/tree/c5fbd55dac29c1fb968a0a1709650def75514167/contrib/seekable_format).
This format is a standard zst file made by many frames follower by a skippable frame that contains the jump table to each frame.
The only drawback is that it's not whidely adopted. The implementation may also have [bugs](https://github.com/facebook/zstd/issues/2335).

Before I discovered this existed I started working on the seek function you requested and ended up with a similar API.
This is the header file of the lib I've been working on:
```
#ifndef _ZSTD_SEEK_
#define _ZSTD_SEEK_

#if defined (__cplusplus)
extern "C" {
#endif

/*
Create a ZSTDSeekContext for the read and seek function.
Returns 0 in case of failure.
uff is a buffer with the whole zstd data, eg a memory mapped file.
size is the size of buff
*/
ZSTDSeekContext* ZSTDSeek_create(void *buff, size_t size);

/*
It reads outBuffSize bytes of uncompressed data from the sctx context buffer into outBuff.
Returns the number of bytes read.
*/
size_t ZSTDSeek_read(void *outBuff, size_t outBuffSize, ZSTDSeekContext *sctx);

/*
Mimic fseek, origin can be SEEK_SET, SEEK_END or SEEK_CUR.
Returns 0 if seek was successfull.
*/
int ZSTDSeek_seek(ZSTDSeekContext *sctx, long offset, int origin);

/*
Free the context.
*/
void ZSTDSeek_free(ZSTDSeekContext *sctx);

#if defined (__cplusplus)
}
#endif

#endif
```

It's working but I have to clean up the code and fix a few corner cases.

The ZSTDSeek_create function will build a jump table in memory while the skippable format API reads it from the file itself.


At this point I have a question for you.

What do you think is the best route to follow?

Should we use the seekable format or should I keep working on my seek function?

If I keep working on my seek function we can seek any zst file, not only those in the seekable format which is rarely used.

The reason it's not popular it's a chicken and egg problem, more details [in the comments of this issue](https://github.com/facebook/zstd/issues/395).

A third option would be for my seek function to recognize if the file uses the seekable format and internally use that API.

**Comment by martinellimarco** [comment]
Oh, I forgot to tell that I've also made a compressor that takes in input a regular .tar file and gives in output a compressed .tar.zst file where each frame starts at the beginning of a tar header and it's composed of a single compressed file.
It's not ready to be published yet but the resulting .tar.zst file should be super fast to navigate once mounted.
I didn't know about the seekable format two days ago when I wrote it but since the only difference is the jumping table at the end I may add one if we go that way.

**Comment by mxmlnkn** [comment]
So, in the end, this format has the exact same problems as gzip and xz have. That's too bad.

If I understood it correctly, the [indexed_gzip](https://github.com/pauldmccarthy/indexed_gzip) module circumvents this by writing out the whole current internal buffer to the index file. I looked a bit at the [zstd RFC](https://tools.ietf.org/html/rfc8478) and it also seems to have such a buffer / window concept, however it can grow almost arbitrarily large, which is 1 KB to 3.75TB, with a recommended maximum of 8MB. Even that recommendation is too large to write out at random points except for really huge files maybe. In comparison the gzip buffer is only [32KB](https://github.com/pauldmccarthy/indexed_gzip/issues/25#issuecomment-576661318), I think. In the first place, it would require a reimplementation or fork, which would be too much work to maintain for me alone.

The problem with multistream zstd is the lack of generality and problems in distinguishing normal zstd files from multistream zstd files. Then again, this definitely wouldn't be the first time. The [Wikipedia dumps](https://en.wikipedia.org/wiki/Wikipedia:Database_download#Should_I_get_multistream?) are stored like this (even though indexed_bzip2 would also work now for normal bz2 files ;)) and [pixz](https://github.com/vasi/pixz) also has the same approach:
> The existing XZ Utils provide great compression in the .xz file format, but they produce just one big block of compressed data. Pixz instead produces a collection of smaller blocks which makes random access to the original data possible. This is especially useful for large tarballs.

The trick with the skippable frame storing a table to each frame is neat, however, for ratarmount it is not necessary because that data could be stored in the ratarmount index file. The zstd decompressor only needs store and load methods for that index. Both, `indexed_gzip` and `indexed_bzip2` have such methods. Therefore, for at least some generality, I'd want to avoid requiring such custom contents in the zstd file.

The problem with the unknown frame size is frustrating. Why even introduce it when it isn't ensured to hold valid values! Then again, bz2 also does not have the decompressed sizes, so you could simply go the naive way and decode each frame once and keep track of the decoded bytes. This part is to some extent the only thing in `indexed_bzip2`, which is different from a standard bz2 decoder.

> What do you think is the best route to follow?

 - Implement seeking support for normal multiframe zstd files. That in turn should also work with that non-standard seekable zstd format even if it ignores the skippable frame. The problem with the unknown frame size can be circumvented by decompressing the frame once and keeping track of its decompressed size ourselves and then store it to the internal index.
 - Is it easy to create that seekable zstd format? Is there a package for that or is it just a proof of concept? If not, then publish your tool for the tar zstd compressor so that multistream/multiframe files can be produced for ratarmount in some way. I would not only do one frame per file but maybe also set a configurable upper size limit for frames maybe with an 8MiB default or so. Think about tars containing super large files in which you want to seek.

> Should we use the seekable format or should I keep working on my seek function?

Personally, I don't see any advantage for that. What that "seekable format" adds is basically the same thing stored in ratarmount's index files. So by using ratarmount's index files, we can keep the original file format untouched. The only advantage for reading that index from the skippable frame at the end instead of computing this table ourselves would be a performance improvement for the very first mounting with ratarmount. But as it would only affect the first time, I don't see it as very important.

**Comment by martinellimarco** [comment]
> Implement seeking support for normal multiframe zstd files. That in turn should also work with that non-standard seekable zstd format even if it ignores the skippable frame. The problem with the unknown frame size can be circumvented by decompressing the frame once and keeping track of its decompressed size ourselves and then store it to the internal index.

I agree, I'll continue my work on this. In the future I may add support to decode existing seekable files transparently in my seek function because the existing seekable format encoder does not seems to fill the uncompressed size field so it's a waste of time to decode the whole frame if I can just read the value from the jump table.

> Is it easy to create that seekable zstd format?

Yes, there is a [seekable_compressor](https://github.com/facebook/zstd/tree/c5fbd55dac29c1fb968a0a1709650def75514167/contrib/seekable_format/examples) in the example folder.

What do you think of the header I've posted?
Do I need to add more methods?
How should I expose the jump table to you? At the moment internally I'm keeping track of the frame index, the compressed position of the start of the frame and the corresponding uncompressed position. Do you need anything else?

**Comment by mxmlnkn** [comment]
> In the future I may add support to decode existing seekable files transparently in my seek function because the existing seekable format encoder does not seems to fill the uncompressed size field so it's a waste of time to decode the whole frame if I can just read the value from the jump table.

Sounds good. That would make the standalone `indexed_zstd` module, or what it may be called, usable for a wider range.

> Yes, there is a [seekable_compressor](https://github.com/facebook/zstd/tree/c5fbd55dac29c1fb968a0a1709650def75514167/contrib/seekable_format/examples) in the example folder.

But is it usable for a normal user? It does not seem like it is available from the zstd Ubuntu package for example. As a user I wouldn't want to compile it myself.

> What do you think of the header I've posted?
> Do I need to add more methods?

I'd need two more functions for querying and setting the frame offsets.
Requiring the whole data in memory seems restricting. I'll have to look into how to memory map a file but I think it should work because the compression is the outermost layer. (That interface wouldn't work for the tar reader because after the transparent decompression I only have a file object at the Python layer not an actual file which I could memory map, or could I memory map "abstract files"?)

> How should I expose the jump table to you? At the moment internally I'm keeping track of the frame index, the compressed position of the start of the frame and the corresponding uncompressed position. Do you need anything else?

That sounds sufficient. `indexed_bzip2` returns this data as a simple map of compressed positions -> uncompressed positions. As you have a C-API, I guess two arrays or one array of structs should be sufficient. For ratarmount it does not matter. `indexed_gzip` basically only returns binary without ratarmount knowing the format. It simply feeds back its own data at a later time.

**Comment by mxmlnkn** [comment]
[Here](https://github.com/mxmlnkn/indexed_bzip2/blob/master/indexed_bzip2/indexed_bzip2.pyx) you can see which functions I really need for the Python interface. For that it would be better to open the decompressor with a simple file name and then also offer close, tell, size. But these are more or less "boiler-plate". Now that I see it, the int returned by your seek is pretty non-standard. Most seek functions I know of, return the new position after the seek thereby emulating a tell to some extent.

**Comment by martinellimarco** [comment]
> But is it usable for a normal user? It does not seem like it is available from the zstd Ubuntu package for example. As a user I wouldn't want to compile it myself.

Good question. I don't think so.

> As you have a C-API, I guess two arrays or one array of structs should be sufficient.

I'm using an array of structs at the moment, I'll expose that.

> Requiring the whole data in memory seems restricting.

Absolutely. That's not the intended behaviour. It is like that beause I started working with ZSTD_decompress that requires the whole frame as a byte array and since I was testing on 10GB compressed archives memory map was the solution.

I've moved to the streaming API but the signature remained because it's a lot easier to work with the memory mapped archive.

If you can use the memory mapped archive good, if not I'll rework the API.

Eventually, what if you mass me the file path and I do the memory mapping internally?

> Here you can see which functions I really need for the Python interface.

Thanks, I'll read that and try to match it as close as possible.

> Now that I see it, the int returned by your seek is pretty non-standard. Most seek functions I know of, return the new position after the seek thereby emulating a tell to some extent.

If that was the case wouldn't the fseek function just return the argument?

I've checked the [C++ reference](https://www.cplusplus.com/reference/cstdio/fseek/) and it states
> If successful, the function returns zero.
Otherwise, it returns non-zero value.

Same on the [GNU C Library](https://www.gnu.org/software/libc/manual/html_node/File-Positioning.html) manual.

**Comment by mxmlnkn** [comment]
> Eventually, what if you mass me the file path and I do the memory mapping internally?

That would be the easiest for me.

> > Now that I see it, the int returned by your seek is pretty non-standard. Most seek functions I know of, return the new position after the seek thereby emulating a tell to some extent.
> 
> If that was the case wouldn't the fseek function just return the argument?
> 
> I've checked the [C++ reference](https://www.cplusplus.com/reference/cstdio/fseek/) and it states
> 
> > If successful, the function returns zero.
> > Otherwise, it returns non-zero value.
> 
> Same on the [GNU C Library](https://www.gnu.org/software/libc/manual/html_node/File-Positioning.html) manual.

Looks like I was thinking of the [Python API](https://docs.python.org/3/library/io.html#io.IOBase.seek) when writing that. But yeah, while in Rome and so on. So leave it as is, returning success. Then, a separate tell function would be necessary. Also, yes, general it returns the offset given in but only when using `SEEK_SET` and then only if the offset is inside the file range. For the relative seeking, it will return absolute offsets, also bounded by the file range.

**Comment by martinellimarco** [comment]
> That would be the easiest for me.

Perfect. I'll do it this way.

> Then, a separate tell function would be necessary.

Ok, no problem.

**Comment by martinellimarco** [comment]
I've published [libzstd-seek](https://github.com/martinellimarco/libzstd-seek) that implements read, seek and tell.

I've tested it and it seems that it works fine. I've also wrote some automated tests using googletest, see [libzstd-seel-tests](https://github.com/martinellimarco/libzstd-seek-tests).

In the example folder there is a small program that list alla file in a .tar.zst file, seeking through each tar header. I've used it to benchmark the library, you'll not need it.

Do you need something else on my part?

**Comment by mxmlnkn** [comment]
One thing I notice is that you are using GPLv3 as a license. I don't think that is "compatible" with my MIT licenses and I don't want to switch to GPLv3. I'm definitely no lawyer but it seems no matter how I use your library, I'd have to switch to GPLv3 even if I only import a GPLv3 Python module [[1]](https://stackoverflow.com/a/999479/2191065) [[2]](https://opensource.stackexchange.com/a/1641/16976) [[3]](https://www.gnu.org/licenses/gpl-faq.html#WhatDoesCompatMean) [[4]](https://www.gnu.org/licenses/gpl-faq.html#GPLStaticVsDynamic) [[5]](https://opensource.stackexchange.com/a/2148/16976) [[6]](https://opensource.stackexchange.com/a/6036/16976) [[7]](https://snyk.io/blog/over-10-of-python-packages-on-pypi-are-distributed-without-any-license/). 
This is insane! It makes me see Python modules in a new light. So, could you please release it under a license without copyleft? I think any of [these](https://snyk.io/blog/over-10-of-python-packages-on-pypi-are-distributed-without-any-license/) is fine:

> The majority of packages on PyPI, 64%, use a “permissive” license, such as the MIT, Apache 2.0 and 2- or 3-Clause BSD licenses.

I would then add libzstd-seek as a git submodule and then build it together with the used zstd library statically to build a manylinux conforming Python package.

Another thing would be about how to distribute the Python module interfacing with your library. For easier distribution, I would want it in an extra Python module. Do you want to own the repository and module or should I just hack together some Python interface code based on `indexed_bzip2` and distribute it myself? I could also add you as a collaborator there. I would also be fine if you copy-pasted the code of `indexed_bzip2` and made the module yourself, it's MIT licensed. I'd even be fine with you opening a repo and me just opening a pull request there so that you don't have to do the "boiler-plate" parts while still owning the repo and all.

As for anything programming related. I think it looks good. If you don't want to do it, I'll try to write the Python interface for it during the weekend. Then, I should be able to tell if something is missing or amiss.

**Comment by martinellimarco** [comment]
Oh, sorry, I didn't think about the license, I usually go with GPLv3 as default. No problem, I'll relicense it to MIT right now.

I have no problem owning / maintaining an `indexed_zstd` repo. Guess it's time to learn python :)
As you say I'll copy-paste `indexed_bzip2` and work from there. If I get stuck I'll ask for your help.

**Comment by mxmlnkn** [comment]
> I have no problem owning / maintaining an `indexed_zstd` repo. Guess it's time to learn python :)
> As you say I'll copy-paste `indexed_bzip2` and work from there. If I get stuck I'll ask for your help.

Oh cool, less work for me, I guess :). If you wanna ask questions per chat instead of here, I can offer:
[![Discord](https://img.shields.io/discord/783411320354766878?label=discord)](https://discord.gg/Wra6t6akh2) [![Telegram](https://img.shields.io/badge/Chat-Telegram-%2330A3E6)](https://t.me/joinchat/FUdXxkXIv6c4Ib8bgaSxNg)
As you can see from the badges, I'll plan to add these to the ReadMe soon as feedback seems to become more common for this project. I'm open to other chat protocols. Unfortunately, I think I'm close to catching them all because everyone wants to use a different chat protocol. Initially, I only planned to add slack but I'm not sure how well it would work. I think it wasn't meant for public usage?

**Comment by martinellimarco** [comment]
I try to reduce the work of the maintainers of the projects to which I contribute.

Besides, the more I do the more I learn, which is always good.

Thanks for the offer, for me Telegram is a good, I'll add you soon.

What's your timezone? I don't want to disturb you more than the bare minimum.

In the meantime, I've published a first version of [indexed_zstd](https://github.com/martinellimarco/indexed_zstd) with 4 FIXME in `ZSTDReader.hpp` that maybe you can help me quickly sort out.

- `ZSTDReader( int fileDescriptor )` do you need a constructor with the fileDescriptor or is it ok to have only the one with filePath and the other with the buffer and its size?
- `fileno`: what should I return here? I don't have the file descriptor exposed. Should I expose it?
- `setBlockOffsets`: is it needed? Altering the jump table is not really possible now
- `tellCompressed`: Is this needed? I can modify my lib to keep track of it, but I don't see the point

Also I'm adding you as a maintainer of this repo.

**Comment by mxmlnkn** [comment]
> What's your timezone? I don't want to disturb you more than the bare minimum.

Good morning :). I'm in CET (UTC+1)

>     * `ZSTDReader( int fileDescriptor )` do you need a constructor with the fileDescriptor or is it ok to have only the one with filePath and the other with the buffer and its size?

I was thinking about this when you asked me about the file open interface before. Opening by file number is not strictly necessary. It's not like I allow piping to ratarmount or plan to do that.

>     * `fileno`: what should I return here? I don't have the file descriptor exposed. Should I expose it?

Couldn't you just return the file descriptor of the opened compressed input? Ratarmount doesn't require it per se. If neither opening by file descriptor nor querying the file descriptor is available, then ratarmount's progress bar during first opening won't work, I think. Actually, indexed_bzip2 is currently the black sheep in ratarmount because [I open it with the fileno](https://github.com/mxmlnkn/ratarmount/blob/develop/ratarmount.py#L1026-L1045) of an already opened file object. For gz and xz, it's more the standard Python way because they are opened using a file object. Might be good to look for inspiration in indexed_gzip.

`setBlockOffsets`: is it needed? Altering the jump table is not really possible now

Yes, it's. On first start, I let the block offsets be created, then I query them with `blockOffsets` and save them out. On subsequent mounts, I then restore them from the saved location and then set them with `setBlockOffsets` in order to avoid parsing the whole file again.

> `tellCompressed`: Is this needed? I can modify my lib to keep track of it, but I don't see the point

It's not needed per se. I added it to mine because I do have the information and it's nice to know. If that method exists, I use it for the progress bar. However, for gzip and xz, I simply call `tell` on the underlying uncompressed file object. However, this only would work
> 
> 
> Also I'm adding you as a maintainer of this repo.

**Comment by martinellimarco** [comment]
Oh, good we are in the same timezone. Now I see  you are from Germany, I'm from Italy.

I now understand better the requirements and I have to refactor libzstd-seek a bit to support `setBlockOffsets` but I think I'll have it done today, it's an easy task.

**Comment by martinellimarco** [comment]
I've implemented all the missing methods and constructors.

While I do now return the internal file descriptor given that it's used in a memory map I'm not sure if you can use it for a progress bar, but `tellCompressed' should do the trick instead.

Is anything missing?

**Comment by mxmlnkn** [comment]
Closed with 87a24a2c1a812fa86d93bdfcb4e20087e021d595
